flight/Libraries/morsel.c

```diff
@@ -0,0 +1,271 @@
+//    and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <morsel.h>
+
+/* Morse stuff is between 1 and 6 elements.
+ * First 2 bits == 11 (punctuation), next 6 bits are character
+ * First 3 bits == 000, 001, 010, 011, 100 for 1 to 5 elements,
+ * following 5 bits are character.
+ *
+ * 101 reserved / unused
+ *
+ * 0 is dit, 1 is dah
+ */
+
+static const uint8_t morse_table[] = {
+	/* **** ASCII, '(' to '@' **** (0..24) vs 40..64 */
+	0B10010110,
+	/*   -.--.	( */
+	0B11101101,
+	/*  -.--.-	) */
+	0B11011010,
+	/*  .--.-.	@ (should be *) */
+	0B10001010,
+	/*   .-.-.	+ */
+	0B11110011,
+	/*  --..--	, */
+	0B11100001,
+	/*  -....-	- */
+	0B11010101,
+	/*  .-.-.-	. */
+	0B10010010,
+	/*   -..-.	/ */
+	0B10011111,
+	/*   -----	0 */
+	0B10001111,
+	/*   .----	1 */
+	0B10000111,
+	/*   ..---	2 */
+	0B10000011,
+	/*   ...--	3 */
+	0B10000001,
+	/*   ....-	4 */
+	0B10000000,
+	/*   .....	5 */
+	0B10010000,
+	/*   -....	6 */
+	0B10011000,
+	/*   --...	7 */
+	0B10011100,
+	/*   ---..	8 */
+	0B10011110,
+	/*   ----.	9 */
+	0B11111000,
+	/*  ---...	: */
+	0B11101010,
+	/*  -.-.-.	; */
+	0B10010110,
+	/*   -.--.	[ (should be <) */
+	0B10010001,
+	/*   -...-	= */
+	0B11101101,
+	/*  -.--.-	] (should be >) */
+	0B11001100,
+	/*  ..--..	? */
+	0B11011010,
+	/*  .--.-.	@ (should be *) */
+
+	/*   **** ASCII, A to Z **** (25..50) vs 65..90 and 97..122 */
+	0B00101000,
+	/*   .-		A */
+	0B01110000,
+	/*   -...	B */
+	0B01110100,
+	/*   -.-.	C */
+	0B01010000,
+	/*   -..	D */
+	0B00000000,
+	/*   .		E */
+	0B01100100,
+	/*   ..-.	F */
+	0B01011000,
+	/*   --.	G */
+	0B01100000,
+	/*   ....	H */
+	0B00100000,
+	/*   ..	I */
+	0B01101110,
+	/*   .---	J */
+	0B01010100,
+	/*   -.-	K */
+	0B01101000,
+	/*   .-..	L */
+	0B00111000,
+	/*   --		M */
+	0B00110000,
+	/*   -.		N */
+	0B01011100,
+	/*   ---	O */
+	0B01101100,
+	/*   .--.	P */
+	0B01111010,
+	/*   --.-	Q */
+	0B01001000,
+	/*   .-.	R */
+	0B01000000,
+	/*   ...	S */
+	0B00010000,
+	/*   -		T */
+	0B01000100,
+	/*   ..-	U */
+	0B01100010,
+	/*   ...-	V SHORT SHORT SHORT LONG */
+	0B01001100,
+	/*   .--	W */
+	0B01110010,
+	/*   -..-	X */
+	0B01110110,
+	/*   -.--	Y */
+	0B01111000,
+	/*   --..	Z */
+};
```

Comment:
Seems to work on OpenLager and have blinked lots of stuff there.  But I doubt I've fully covered it.